test/unit/puppet_fact_parser_test.rb

```diff
@@ -233,8 +234,35 @@ def setup
     assert_equal '172.30.43.0', parser.interfaces['local_area_connection_2'][:network]
   end
 
+  test "should create connection between new os and new arch on facts parsing" do
+    assert os_to_arch_connectivity_check
+  end
+
+  test "should create connection between existing os and new arch on facts parsing" do
+    FactoryGirl.create(:operatingsystem,:name => 'AnyOS', :major => '6')
+    assert os_to_arch_connectivity_check
```

Comment:
I think it's not a responsibility of host nor fact parser. Ideally we'd have some host factory that would create and associate everything correctly while parser would just provide hashes of attributes. Since we're not there yet I agree with @domcleal, that it should go to [host](https://github.com/theforeman/foreman/blob/develop/app/models/host/base.rb#L137) model. Or at least if we don't agree on it, I'd suggest making it explicit method of parser since I wouldn't expect `PuppetFactParser#operatingsystem` to create any association. Ideally a method in `FactParser` (it's not specific to puppet) so this way all other parsers would use the same code.